compressors/arithmetic_coding.py

```diff
@@ -0,0 +1,273 @@
+            # and after a few iterations things will be infeasible.
+            # The goal of re-normalizing is to not let the range (high - low) get smaller than self.QTR
+
+            # CASE I, II -> simple cases where low, high are both in the same half
+            while (high < self.HALF) or (low > self.HALF):
+                if high < self.HALF:
+                    # output 1's corresponding to prior mid-range readjustments
+                    encoded_bitarray.extend("0" + "1" * num_mid_range_readjust)
+
+                    # re-adjust range, and reset params
+                    low = low << 1
+                    high = high << 1
+                    num_mid_range_readjust = 0  # reset the mid-range readjustment counter
+
+                elif low > self.HALF:
+                    # output 0's corresponding to prior mid-range readjustments
+                    encoded_bitarray.extend("1" + "0" * num_mid_range_readjust)
+
+                    # re-adjust range, and reset params
+                    low = (low - self.HALF) << 1
+                    high = (high - self.HALF) << 1
+                    num_mid_range_readjust = 0  # reset the mid-range readjustment counter
+
+            # CASE III -> the more complex case where low, high straddle the midpoint
+            while (low > self.QTR) and (high < 3 * self.QTR):
+                # increment the mid-range adjustment counter
+                num_mid_range_readjust += 1
+                low = (low - self.QTR) << 1
+                high = (high - self.QTR) << 1
+
+        # Finally output a few bits to signal the final range + any remaining mid range readjustments
+        num_mid_range_readjust += 1  # this increment is mainly to output either 01, 10
+        if low <= self.QTR:
+            # output 0's corresponding to prior mid-range readjustments
+            encoded_bitarray.extend("0" + num_mid_range_readjust * "1")
+        else:
+            # output 1's corresponding to prior mid-range readjustments
+            encoded_bitarray.extend("1" + num_mid_range_readjust * "0")
+
+        return encoded_bitarray
+
+
+class ArithmeticDecoder(DataDecoder):
+    """Finite precision Arithmetic decoder
+
+    The encoder are decoders are based on the following sources:
+    - https://youtu.be/ouYV3rBtrTI: This series of videos on Arithmetic coding are a very gradual but a great
+    way to understand them
+    - Charles Bloom's blog: https://www.cbloom.com/algs/statisti.html#A5
+    - There is of course the original paper: ADD LINK
+    """
+
+    def __init__(self, precision, freqs):
+        super().__init__()
+        self.freqs = freqs
+
+        # params
+        self.DATA_BLOCK_SIZE_BITS = 32  # represents the size of the data block
+        self.PRECISION = precision
+        self.FULL = 1 << (precision)
+        self.HALF = 1 << (precision - 1)
+        self.QTR = 1 << (precision - 2)
+
+    def shrink_range(self, s: Any, low: int, high: int) -> Tuple[int, int]:
```

Comment:
is this same as the encoder function? If yes, can combine into one by creating a helper class

Reply:
yes, there is lot of overlap. I will try to clean this up later today

Reply:
Done